File: backend/src/Invasion/Server/Protocol.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TemplateHaskell #-}

-- | Wire types for the lobby and per-game WebSockets.
--
-- The frontend mirror lives in @frontend/src/api/protocol.ts@; keep the
-- two in lockstep. Aeson's default 'TaggedObject' encoding produces
-- @{ "tag": "ConstructorName", ... }@ which the TS side discriminates on.
module Invasion.Server.Protocol
  ( -- * Common
    UserInfo (..)
  , ChatLine (..)
  , Visibility (..)
  , GameStatus (..)
  , GameSummary (..)
  , SeatView (..)
  , DeckView (..)
  , GameView (..)
    -- * Lobby socket
  , LobbyIn (..)
  , LobbyOut (..)
    -- * Game socket
  , GameIn (..)
  , GameOut (..)
  , ZoneTarget (..)
  ) where

import Data.Aeson (FromJSON, ToJSON, Value)
import Data.Aeson.TH (deriveJSON, defaultOptions)
import Data.Text (Text)
import Data.Time (UTCTime)
import Data.UUID (UUID)
import GHC.Generics (Generic)
import Invasion.Prelude
import Invasion.Types (CardCode, PlayerKey, UnitKey, ZoneKind)

-- ----------------------------------------------------------------------------
-- Common

data UserInfo = UserInfo
  { userId :: UUID
  , displayName :: Text
  }
  deriving stock (Show, Eq, Generic)

data ChatLine = ChatLine
  { from :: UserInfo
  , text :: Text
  , at :: UTCTime
  }
  deriving stock (Show, Generic)

data Visibility = Public | Private
  deriving stock (Show, Eq, Generic)

data GameStatus
  = StatusWaiting
  | StatusPlaying
  | StatusEnded
  deriving stock (Show, Eq, Generic)

data DeckView = DeckView
  { deckId :: UUID
  , name :: Text
  , capital :: Maybe Text
  , size :: Int
  }
  deriving stock (Show, Generic)

data SeatView = SeatView
  { seat :: Text -- "Player1" | "Player2"
  , user :: UserInfo
  , isHost :: Bool
  , deck :: Maybe DeckView
  }
  deriving stock (Show, Generic)

-- | What lobby clients see in the game list. Private games are not
-- broadcast at all, so this only ever describes public ones.
data GameSummary = GameSummary
  { gameId :: UUID
  , name :: Text
  , host :: UserInfo
  , visibility :: Visibility
  , hasPassword :: Bool
  , filledSeats :: Int -- 0..2
  , status :: GameStatus
  , allowSpectators :: Bool
  , spectatorCount :: Int
  }
  deriving stock (Show, Generic)

-- | The full server-side view of a game, sent to clients connected to
-- its game socket. 'inviteToken' is only populated for the host.
data GameView = GameView
  { gameId :: UUID
  , name :: Text
  , host :: UserInfo
  , visibility :: Visibility
  , hasPassword :: Bool
  , allowSpectators :: Bool
  , spectatorCount :: Int
  , inviteToken :: Maybe Text
  , seats :: [SeatView]
  , status :: GameStatus
  , chat :: [ChatLine]
  , engine :: Maybe Value
    -- ^ The engine 'Game' value, serialized as a JSON blob, present once
    -- the game has been started. Sent as opaque 'Value' here so this
    -- module doesn't need to import the engine; the frontend has a
    -- typed view of the relevant subset. Eventually this becomes a
    -- per-viewer snapshot (opponent's hand hidden).
  }
  deriving stock (Show, Generic)

-- ----------------------------------------------------------------------------
-- Lobby socket

data LobbyIn
  = -- | Send a chat line to the global lobby.
    LobbyChatSend { text :: Text }
  | -- | Create a new game slot. Server replies with 'LobbyGameCreated'.
    -- 'allowSpectators' is optional: if absent, the server defaults to
    -- True for public games and False for private ones.
    LobbyCreateGame
      { name :: Text
      , visibility :: Visibility
      , password :: Maybe Text
      , allowSpectators :: Maybe Bool
      }
  | -- | Join a public game (no password needed). Reply: 'LobbyGameJoinOk'.
    LobbyJoinPublic { gameId :: UUID }
  | -- | Join a password-gated game. Reply: 'LobbyGameJoinOk' or 'LobbyError'.
    LobbyJoinWithPassword { gameId :: UUID, password :: Maybe Text }
  deriving stock (Show, Generic)

data LobbyOut
  = LobbyWelcome
      { you :: UserInfo
      , users :: [UserInfo]
      , games :: [GameSummary]
      , chat :: [ChatLine]
      }
  | LobbyChatNew { line :: ChatLine }
  | LobbyUsersUpdate { users :: [UserInfo] }
  | LobbyGamesUpdate { games :: [GameSummary] }
  | LobbyGameCreated
      { gameId :: UUID
      , inviteToken :: Maybe Text
      }
  | LobbyGameJoinOk
      { gameId :: UUID
      , inviteToken :: Maybe Text
      }
  | LobbyError { code :: Text }
  deriving stock (Show, Generic)

-- ----------------------------------------------------------------------------
-- Per-game socket

data GameIn
  = GameChatSend { text :: Text }
  | -- | Replace this seat's loaded deck with the deck identified by id.
    -- The deck must belong to the seated user.
    GameSelectDeck { deckId :: UUID }
  | -- | Clear the loaded deck for this seat.
    GameClearDeck
  | -- | Host-only. Transition Waiting -> Playing if both seats have decks.
    GameStart
  | -- | Pass the current action window. The server fills in the
    -- 'PlayerKey' from the sender's seat; if it isn't this player's
    -- priority the engine will silently ignore the message.
    GamePassPriority
  | -- | Play a card from the sender's hand. The card is identified by
    -- its stable 'UnitKey' (the same key the frontend sees on the card
    -- it clicked) so duplicates in hand are addressed unambiguously.
    --
    --   * Unit / Support (non-attachment): 'zone' picks which zone the
    --     card enters. Required for those kinds; ignored otherwise.
    --   * Support (attachment trait): 'target' picks the host unit's
    --     'UnitKey'. Required for attachments; ignored otherwise.
    --   * Quest / Tactic / Legend: neither 'zone' nor 'target' is read.
    GamePlayCard
      { cardKey :: UnitKey
      , zone :: Maybe ZoneKind
      , target :: Maybe UnitKey
      }
  | -- | Trigger a printed action ability on an in-play card. The
    -- engine validates that the source belongs to the sender, debits
    -- the resource cost, and checks the supplied target against the
    -- action's declared 'TargetSchema'.
    GameTriggerAction
      { source :: UnitKey
      , actionIndex :: Int
      , target :: Maybe UnitKey
      , targetZone :: Maybe ZoneTarget
      }
  | -- | Drop this user from the seat, broadcast to the other seat.
    GameLeave
  deriving stock (Show, Generic)

-- | Zone reference for action targets that point at a capital zone.
data ZoneTarget = ZoneTarget
  { player :: PlayerKey
  , kind :: ZoneKind
  }
  deriving stock (Show, Generic)

data GameOut
  = GameWelcome
      { you :: UserInfo
      , game :: GameView
      }
  | GameUpdate { game :: GameView }
  | GameChatNew { line :: ChatLine }
  | GameError { code :: Text }
  | -- | Sent when the slot is being torn down. Frontend should redirect
    -- back to the lobby.
    GameClosed { reason :: Text }
  deriving stock (Show, Generic)

mconcat
  [ deriveJSON defaultOptions ''UserInfo
  , deriveJSON defaultOptions ''ChatLine
  , deriveJSON defaultOptions ''Visibility
  , deriveJSON defaultOptions ''GameStatus
  , deriveJSON defaultOptions ''DeckView
  , deriveJSON defaultOptions ''SeatView
  , deriveJSON defaultOptions ''GameSummary
  , deriveJSON defaultOptions ''GameView
  , deriveJSON defaultOptions ''LobbyIn
  , deriveJSON defaultOptions ''LobbyOut
  , deriveJSON defaultOptions ''ZoneTarget
  , deriveJSON defaultOptions ''GameIn
  , deriveJSON defaultOptions ''GameOut
  ]

-- Pacify -Wunused-top-binds on the generic-derived hooks above.
_unused :: (FromJSON UserInfo, ToJSON UserInfo) => ()
_unused = ()

File: frontend/src/api/protocol.ts
// Wire types for the lobby and per-game WebSockets.
// Mirror of backend/src/Invasion/Server/Protocol.hs. When that file
// changes, this one changes in the same PR.

export interface UserInfo {
  userId: string
  displayName: string
}

export interface ChatLine {
  from: UserInfo
  text: string
  at: string // ISO timestamp from the server
}

export type Visibility = 'Public' | 'Private'

export type GameStatus = 'StatusWaiting' | 'StatusPlaying' | 'StatusEnded'

export interface DeckView {
  deckId: string
  name: string
  capital: string | null
  size: number
}

export interface SeatView {
  seat: string // "Player1" | "Player2"
  user: UserInfo
  isHost: boolean
  deck: DeckView | null
}

export interface GameSummary {
  gameId: string
  name: string
  host: UserInfo
  visibility: Visibility
  hasPassword: boolean
  filledSeats: number
  status: GameStatus
}

export interface GameView {
  gameId: string
  name: string
  host: UserInfo
  visibility: Visibility
  hasPassword: boolean
  inviteToken: string | null
  seats: SeatView[]
  status: GameStatus
  chat: ChatLine[]
  // The engine snapshot. `null` while the game is waiting; populated
  // once Setup + BeginGame have run. Kept here as a single payload so
  // the renderer can derive everything from one source of truth.
  engine: EngineGame | null
}

// ---------------------------------------------------------------------------
// Engine snapshot
// Mirror of the JSON produced by Invasion.Game's ToJSON instance.

export type PlayerKey = 'Player1' | 'Player2'

export type Phase = 'KingdomPhase' | 'QuestPhase' | 'CapitalPhase' | 'BattlefieldPhase'

export type ZoneKind = 'KingdomZone' | 'QuestZone' | 'BattlefieldZone'

export interface EngineZone {
  kind: ZoneKind
  developments: number
  damage: number
  burned: boolean
  // hitPoints is derived (8 + developments); not on the wire.
}

export interface EngineCapital {
  kingdom: EngineZone
  quest: EngineZone
  battlefield: EngineZone
}

export type ActionWindowTrigger =
  | 'KingdomActionWindow'
  | 'QuestActionWindow'
  | 'CapitalActionWindow'
  | 'BattlefieldActionWindow'
  | 'AfterDeclareCombatTarget'
  | 'AfterDeclareAttackers'
  | 'AfterDeclareDefenders'
  | 'AfterAssignCombatDamage'
  | 'AfterApplyCombatDamage'

export type PassState =
  | { tag: 'NoPasses'; contents: PlayerKey }
  | { tag: 'OnePass'; contents: PlayerKey }

export interface EngineActionWindow {
  trigger: ActionWindowTrigger
  awaiting: PassState
}

export type EliminationReason = 'DeckedOut' | 'CapitalBurned'

export type PlayerLifecycle =
  | { tag: 'IdlePlayer' }
  | { tag: 'Eliminated'; contents: EliminationReason }
  | { tag: 'PlayerDraw'; contents: unknown }

// Currently only Dwarf exists engine-side, but the asset set covers the
// full Warhammer: Invasion race list. Add to this union as we add races
// to backend/src/Invasion/Types.hs.
export type Race =
  | 'Dwarf'
  | 'Empire'
  | 'HighElf'
  | 'Chaos'
  | 'Orc'
  | 'DarkElf'

export interface EnginePlayer {
  key: PlayerKey
  state: PlayerLifecycle
  capital: EngineCapital
  resources: number
  // hand / deck / discard are arrays of card-def JSON; we only need
  // their .length for the basic in-game UI, so leave them unknown.
  hand: unknown[]
  deck: unknown[]
  discard: unknown[]
  race: Race
}

export type GameLifecycle =
  | { tag: 'GameSetup' }
  | { tag: 'GamePlaying' }
  | {
      tag: 'GameFinished'
      contents: {
        winner: PlayerKey
        reason: 'OpponentDeckedOut' | 'OpponentCapitalBurned'
      }
    }

export type LogCategory =
  | 'LogSystem'
  | 'LogPhase'
  | 'LogTurn'
  | 'LogPlayerAction'
  | 'LogResult'

// Engine-emitted transcript entry. `key` is an i18n key (resolved in
// frontend/src/locales/). `params` carries interpolation values; enum-
// shaped values (player keys, phases, triggers, reasons) are written
// raw and resolved through nested i18n lookups before substitution —
// see `formatLogEntry` in the game view.
export interface LogEntry {
  at: string
  category: LogCategory
  key: string
  params: Record<string, string>
}

export interface EngineGame {
  player1: EnginePlayer
  player2: EnginePlayer
  firstPlayer: PlayerKey
  currentPlayer: PlayerKey
  turn: number
  phase: Phase | null
  actionWindow: EngineActionWindow | null
  modifiers: unknown
  lifecycle: GameLifecycle
  log: LogEntry[]
  units: unknown[]
  nextUnitKey: unknown
}

// Derived helpers — keep alongside the wire types so they stay in sync.
export function zoneHitPoints(z: EngineZone): number {
  return 8 + z.developments
}

export function zoneBurning(z: EngineZone): boolean {
  return z.damage >= zoneHitPoints(z)
}

export function priorityHolder(s: PassState): PlayerKey {
  return s.contents
}

// ---------------------------------------------------------------------------
// Lobby socket

export type LobbyIn =
  | { tag: 'LobbyChatSend'; text: string }
  | {
      tag: 'LobbyCreateGame'
      name: string
      visibility: Visibility
      password: string | null
    }
  | { tag: 'LobbyJoinPublic'; gameId: string }
  | { tag: 'LobbyJoinWithPassword'; gameId: string; password: string | null }

export type LobbyOut =
  | {
      tag: 'LobbyWelcome'
      you: UserInfo
      users: UserInfo[]
      games: GameSummary[]
      chat: ChatLine[]
    }
  | { tag: 'LobbyChatNew'; line: ChatLine }
  | { tag: 'LobbyUsersUpdate'; users: UserInfo[] }
  | { tag: 'LobbyGamesUpdate'; games: GameSummary[] }
  | { tag: 'LobbyGameCreated'; gameId: string; inviteToken: string | null }
  | { tag: 'LobbyGameJoinOk'; gameId: string; inviteToken: string | null }
  | { tag: 'LobbyError'; code: string }

// ---------------------------------------------------------------------------
// Game socket

export type GameIn =
  | { tag: 'GameChatSend'; text: string }
  | { tag: 'GameSelectDeck'; deckId: string }
  | { tag: 'GameClearDeck' }
  | { tag: 'GameStart' }
  | { tag: 'GamePassPriority' }
  | { tag: 'GameLeave' }

export type GameOut =
  | { tag: 'GameWelcome'; you: UserInfo; game: GameView }
  | { tag: 'GameUpdate'; game: GameView }
  | { tag: 'GameChatNew'; line: ChatLine }
  | { tag: 'GameError'; code: string }
  | { tag: 'GameClosed'; reason: string }
